x-pack/functionbeat/provider/gcp/gcp/pubsub.go

```diff
@@ -0,0 +1,106 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package gcp
+
+import (
+	"context"
+	"fmt"
+
+	"cloud.google.com/go/pubsub"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/feature"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/x-pack/functionbeat/function/core"
+	"github.com/elastic/beats/x-pack/functionbeat/function/provider"
+	"github.com/elastic/beats/x-pack/functionbeat/provider/gcp/gcp/transformer"
+)
+
+const (
+	pubSubEventCtxStr = "pub_sub_event"
+)
+
+// PubSub represents a Google Cloud function which reads event from Google Pub/Sub triggers.
+type PubSub struct {
+	log    *logp.Logger
+	config *FunctionConfig
+}
+
+// PubSubEventKey is an alias to string
+type PubSubEventKey string
+
+// NewPubSub returns a new function to read from Google Pub/Sub.
+func NewPubSub(provider provider.Provider, cfg *common.Config) (provider.Function, error) {
+	config := defaultPubSubFunctionConfig()
+	err := cfg.Unpack(config)
+	if err != nil {
+		return &PubSub{}, err
+	}
+
+	return &PubSub{
+		log:    logp.NewLogger("pubsub"),
+		config: config,
+	}, nil
+}
+
+// PubSubEvent stores the context and the message from Google Pub/Sub.
+type PubSubEvent struct {
+	Ctx     context.Context
```

Comment:
Uhm... why do we have to wrap a `context.Context` in a custom struct that is passed via `context.Context` as well?

Reply:
Extracted `*metadata.Metadata` from Context in the current implementation.